src/Network/HTTPClient.php

```diff
@@ -0,0 +1,253 @@
+use Psr\Log\InvalidArgumentException;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Performs HTTP requests to a given URL
+ */
+class HTTPClient implements IHTTPClient
+{
+	/** @var LoggerInterface */
+	private $logger;
+	/** @var Profiler */
+	private $profiler;
+	/** @var Client */
+	private $client;
+	/** @var URLResolver */
+	private $resolver;
+
+	public function __construct(LoggerInterface $logger, Profiler $profiler, Client $client, URLResolver $resolver)
+	{
+		$this->logger   = $logger;
+		$this->profiler = $profiler;
+		$this->client   = $client;
+		$this->resolver = $resolver;
+	}
+
+	/**
+	 * @throws HTTPException\InternalServerErrorException
+	 */
+	protected function request(string $method, string $url, array $opts = []): IHTTPResult
+	{
+		$this->profiler->startRecording('network');
+		$this->logger->debug('Request start.', ['url' => $url, 'method' => $method]);
+
+		if (Network::isLocalLink($url)) {
+			$this->logger->info('Local link', ['url' => $url, 'callstack' => System::callstack(20)]);
+		}
+
+		if (strlen($url) > 1000) {
+			$this->logger->debug('URL is longer than 1000 characters.', ['url' => $url, 'callstack' => System::callstack(20)]);
+			$this->profiler->stopRecording();
+			return CurlResult::createErrorCurl(substr($url, 0, 200));
+		}
+
+		$parts2     = [];
+		$parts      = parse_url($url);
+		$path_parts = explode('/', $parts['path'] ?? '');
+		foreach ($path_parts as $part) {
+			if (strlen($part) <> mb_strlen($part)) {
+				$parts2[] = rawurlencode($part);
+			} else {
+				$parts2[] = $part;
+			}
+		}
+		$parts['path'] = implode('/', $parts2);
+		$url           = Network::unparseURL($parts);
+
+		if (Network::isUrlBlocked($url)) {
+			$this->logger->info('Domain is blocked.', ['url' => $url]);
+			$this->profiler->stopRecording();
+			return CurlResult::createErrorCurl($url);
+		}
+
+		$conf = [];
+
+		if (!empty($opts['cookiejar'])) {
+			$jar                           = new FileCookieJar($opts['cookiejar']);
+			$conf[RequestOptions::COOKIES] = $jar;
+		}
+
+		$header = [];
+
+		if (!empty($opts['accept_content'])) {
+			$header['Accept'] = $opts['accept_content'];
+		}
+
+		if (!empty($opts['header'])) {
+			$header = array_merge($opts['header'], $header);
+		}
+
+		if (!empty($opts['headers'])) {
+			$this->logger->notice('Wrong option \'headers\' used.');
+			$header = array_merge($opts['headers'], $header);
+		}
+
+		$conf[RequestOptions::HEADERS] = array_merge($this->client->getConfig(RequestOptions::HEADERS), $header);
+
+		if (!empty($opts['timeout'])) {
+			$conf[RequestOptions::TIMEOUT] = $opts['timeout'];
+		}
+
+		$conf[RequestOptions::ON_HEADERS] = function (ResponseInterface $response) use ($opts) {
+			if (!empty($opts['content_length']) &&
+				$response->getHeaderLine('Content-Length') > $opts['content_length']) {
+				throw new TransferException('The file is too big!');
+			}
+		};
+
+		try {
+			switch ($method) {
+				case 'get':
+				case 'head':
+				case 'post':
```

Comment:
PUT and DELETE are pretty popular too.
```suggestion
				case 'get':
				case 'head':
				case 'post':
				case 'put':
				case 'delete':
```